src/runtime/components/Sections/index.vue

```diff
@@ -1931,6 +2092,164 @@ const selectedCSS = (mediaObject, mediaFieldName) => {
 const removeMedia = (media) => {
   pageMetadata.value[media] = {}
 }
+const unsavedSettings = (tab) => {
+  if (tab === 'page_settings') {
+    try {
+      let originalSettings = {}
+      locales.value.forEach(locale => {
+        originalSettings[locale] = originalMetaData.value[locale]
+      })
+      originalSettings.pagePath = originalMetaData.value.pagePath
+      originalSettings.media = originalMetaData.value.media
+      originalSettings.mediaMetatag = originalMetaData.value.mediaMetatag
+      if (!originalSettings.media) {
+        originalSettings.media = {}
+      }
+      if (!originalSettings.mediaMetatag) {
+        originalSettings.mediaMetatag = {}
+      }
+      let updatedSettings = {}
+      updatedSettings = {
+        ...pageMetadata.value,
+        pagePath: pagePath.value
+      }
+      if (!updatedSettings.media) {
+        updatedSettings.media = {}
+      }
+      if (!updatedSettings.mediaMetatag) {
+        updatedSettings.mediaMetatag = {}
+      }
+
+      unsavedSettingsError.value[tab] = !isEqual(originalSettings, updatedSettings)
+
+      return unsavedSettingsError.value[tab]
+    } catch {
+      unsavedSettingsError.value[tab] = false
+      return false
+    }
+  } else if (tab === 'builder_settings') {
+    try {
+      const allEmpty = Object.values(updatedBuilderSettings.value).every(value => value.trim() === '')
+      unsavedSettingsError.value[tab] = !isEqual(originalBuilderSettings.value, updatedBuilderSettings.value) && !allEmpty
+
+      return unsavedSettingsError.value[tab]
+    } catch {
+      unsavedSettingsError.value[tab] = false
+      return false
+    }
+  }
+}
+const getBuilderComponentForm = (component_name) => {
+  const path = `/builder/${component_name}`;
+  return importComp(path).component;
+};
+const switchSettingsTab = (tab) => {
+  unsavedSettings(currentPageSettingsTab.value)
+  currentPageSettingsTab.value = tab
+}
+const overwriteCssStyleRoot = (settings) => {
+  const cssVars = Object.entries(settings)
+    .filter(([, value]) => value != null && value !== '') // exclude null, undefined, and empty string
+    .map(([property, value]) => `  ${property}: ${value} !important;`)
+    .join('\n')
+
+  computedInjectedStyle.value = `:root {\n${cssVars}\n}`
+}
+const applySettingsToCSS = (settings) => {
+  const root = document.documentElement
+
+  // Apply each setting as a CSS custom property
+  Object.entries(settings).forEach(([property, value]) => {
+    // Set the CSS custom property with !important to override existing values
+    root.style.setProperty(property, value, 'important')
+  })
+}
+const builderSettingUpdated = (settings) => {
+  if (!projectMetadata.value.builder) {
+    projectMetadata.value.builder = {
+      builder_settings: {}
+    }
+  } else if (!projectMetadata.value.builder.builder_settings) {
+    projectMetadata.value.builder.builder_settings = {}
+  }
+  projectMetadata.value.builder.builder_settings = settings
+  applySettingsToCSS(settings)
```

Comment:
should be called by the builder not the library, cleanly separate concerns here...

Reply:
looks better